promql/test.go

```diff
@@ -416,6 +450,171 @@ func (cmd *loadCmd) append(a storage.Appender) error {
+func (cmd *loadCmd) appendCustomHistogram(a storage.Appender) error {
+	histMap := map[uint64]tempHistogramWrapper{}
+
+	// Go through all the time series to collate classic histogram data
+	// and organise them by timestamp.
+	for hash, smpls := range cmd.defs {
+		m := cmd.metrics[hash]
+		mName := m.Get(labels.MetricName)
+		switch {
+		case strings.HasSuffix(mName, "_bucket") && m.Has(labels.BucketLabel):
+			le, err := strconv.ParseFloat(m.Get(labels.BucketLabel), 64)
+			if err != nil {
+				continue
+			}
+			processClassicHistogramSeries(m, "_bucket", histMap, smpls, func(histWrapper *tempHistogramWrapper) {
+				histWrapper.upperBounds = append(histWrapper.upperBounds, le)
+			}, func(hist *tempHistogram, f float64) {
+				hist.bucketCounts[le] = f
+			})
+		case strings.HasSuffix(mName, "_count"):
+			processClassicHistogramSeries(m, "_count", histMap, smpls, nil, func(hist *tempHistogram, f float64) {
+				hist.count = f
+			})
+		case strings.HasSuffix(mName, "_sum"):
+			processClassicHistogramSeries(m, "_sum", histMap, smpls, nil, func(hist *tempHistogram, f float64) {
+				hist.sum = f
+			})
+		}
+	}
+
+	// Convert the collated classic histogram data into native histograms
+	// with custom bounds and append them to the storage.
+	for _, histWrapper := range histMap {
+		sort.Float64s(histWrapper.upperBounds)
+		upperBounds := make([]float64, 0, len(histWrapper.upperBounds))
+		prevLe := -372.4869 // assuming that the lowest bound is never this value
+		for _, le := range histWrapper.upperBounds {
+			if le != prevLe { // deduplicate
+				upperBounds = append(upperBounds, le)
+				prevLe = le
+			}
+		}
+		var customBounds []float64
+		if upperBounds[len(upperBounds)-1] == math.Inf(1) {
+			customBounds = upperBounds[:len(upperBounds)-1]
+		} else {
+			customBounds = upperBounds
+		}
+		fhBase := &histogram.FloatHistogram{
+			Count:  0,
+			Sum:    0,
+			Schema: histogram.CustomBucketsSchema,
+			PositiveSpans: []histogram.Span{
+				{Offset: 0, Length: uint32(len(upperBounds))},
+			},
+			CustomValues: customBounds,
+		}
+		ts := make([]int64, 0, len(histWrapper.histByTs))
+		for t := range histWrapper.histByTs {
+			ts = append(ts, t)
+		}
+		sort.Slice(ts, func(i, j int) bool { return ts[i] < ts[j] })
+		for _, t := range ts {
+			hist, exists := histWrapper.histByTs[t]
+			if !exists {
+				continue
+			}
+			fh := fhBase.Copy()
+			counts := make([]float64, 0, len(upperBounds))
+			var prevCount, total float64
+			for _, le := range upperBounds {
+				currCount, exists := hist.bucketCounts[le]
+				if !exists {
+					currCount = 0
+				}
+				count := currCount - prevCount
+				counts = append(counts, count)
+				total += count
+				prevCount = currCount
+			}
+			fh.Sum = hist.sum
+			if hist.count != 0 {
+				total = hist.count
+			}
+			fh.Count = total
+			fh.PositiveBuckets = counts
+			// s := Sample{Metric: m, T: t, H: fh}
+			// fmt.Printf("end s: %v\n", s)
+			s := Sample{T: t, H: fh}
```

Comment:
Should probably call Compact() as discussed above.

```suggestion
			s := Sample{T: t, H: fh.Compact()}
```